monitoring/uss_qualifier/suites/suite.py

```diff
@@ -0,0 +1,198 @@
+from datetime import datetime
+from enum import Enum
+import json
+import os
+from typing import Dict, List, Optional
+
+from implicitdict import ImplicitDict, StringBasedDateTime
+import yaml
+
+from monitoring.monitorlib.inspection import fullname
+from monitoring.uss_qualifier.reports import (
+    TestScenarioReport,
+    FailedCheck,
+    TestSuiteReport,
+    TestSuiteActionReport,
+)
+from monitoring.uss_qualifier.resources import Resource, ResourceID, ResourceType
+from monitoring.uss_qualifier.scenarios.scenario import (
+    TestScenario,
+    TestScenarioDeclaration,
+)
+
+
+SuiteType = str
+"""This plain string represents a type of test suite, expressed as the file name of the suite definition (without extension) qualified relative to this `suites` folder"""
+
+
+class TestSuiteDeclaration(ImplicitDict):
+    suite_type: SuiteType
+    """Type of test suite"""
+
+    resources: Dict[ResourceID, ResourceID]
+    """Mapping of the ID a resource will be known by in the child test suite -> the ID a resource is known by in the parent test suite.
+    
+    The child suite resource <key> is supplied by the parent suite resource <value>.
+    """
+
+
+class ReactionToFailure(str, Enum):
+    Continue = "Continue"
+    """If the test suite action fails, continue to the next action in that test suite"""
+
+    Abort = "Abort"
+    """If the test suite action fails, do not execute any more actions in that test suite"""
+
+
+class TestSuiteActionDeclaration(ImplicitDict):
+    """Defines a step in the sequence of things to do for a test suite.
+
+    Exactly one of `test_scenario` or `test_suite` must be specified.
+    """
+
+    test_scenario: Optional[TestScenarioDeclaration]
+    """If this field is populated, declaration of the test scenario to run"""
+
+    test_suite: Optional[TestSuiteDeclaration]
+    """If this field is populated, declaration of the test suite to run"""
+
+    on_failure: ReactionToFailure
+    """What to do if this action fails"""
+
+
+class TestSuiteDefinition(ImplicitDict):
+    """Schema for the definition of a test suite, analogous to the Python TestScenario subclass for scenarios"""
+
+    name: str
+    """Name of the test suite"""
+
+    resources: Dict[ResourceID, ResourceType]
+    """Enumeration of the resources used by this test suite"""
+
+    actions: List[TestSuiteActionDeclaration]
+    """The actions to take when running the test suite.  Components will be executed in order."""
+
+    @staticmethod
+    def load(suite_type: str) -> "TestSuiteDefinition":
```

Comment:
`TestSuiteDefinition` type is between quotes. Is it wanted?

Reply:
Yes, it essentially allows forward declaration of types.  If this type wasn't in quotes and we don't have `from __future__ import annotations`, then Python will have an error since TestSuiteDefinition is not defined at this point.  Another way to address the issue is to include that __future__ statement, but I've seen a couple cases where that changed ImplicitDict behavior in unexpected ways and I didn't want to get that all ironed out at the moment.

Reply:
Nice, thank you for the explanation.